### ISSUE TITLE
`open type` does not bring enum cases into scope in patterns

### ISSUE THREAD

**Issue by brianrourkeboll** [dotnet/fsharp#17630]
I meant to open this years ago, but I guess I never did.

**Repro steps**

```fsharp
type E = A = 1 | B = 2

open type E

let f x =
    match x with
    | A -> () // This binds a new value `A` instead of matching the enum case E.A, which is otherwise in scope unqualified.
```

**Expected behavior**

Enum cases brought into scope with an `open type` declaration (or an `AutoOpen` on the enum definition) should be in scope in pattern matching as unqualified literals, just like union cases or active pattern cases.

E.g., if you change the enum in the example above to a union type, `A` _is_ brought into scope:

```fsharp
type E = A | B

open type E

let f x =
    match x with
    | A -> ()
    | B -> ()
```

**Actual behavior**

Enum cases brought into scope with an `open type` declaration (or an `AutoOpen` on the enum definition) are not in scope in pattern matching as unqualified literals.

**Known workarounds**

N/A.

**Related information**

.NET 8, 9.

**Note**

I guess this probably can't be fixed, since there may now be existing code that relies on the current behavior — although it is quite possible that some of that existing code is wrong, because it was written with the understanding that `open type` would behave the same for enums as it does for unions, etc.

**Comment by edgarfgp** [comment]
`open type` should work the same for enum and unions. I propose we add a new warning(Informational) and adjust the current analysis to do the right thing.

**Comment by abonie** [comment]
Weirdly enough the `open type` with enums behaves like this only in pattern matches it seems.

Technically a breaking change so we should at least add a warning together with a fix.

**Comment by auduchinok** [comment]
Enum types are considered to have `RequireQualifiedAccess`, so I'd say it's an error that it's possible to open an enum type at all.

**Comment by brianrourkeboll** [comment]
> Enum types are considered to have `RequireQualifiedAccess`

They are? I mean, I guess they were before `open type`, because they "required" qualified access by default (unlike F# unions), and there was no way to open them. But that is true of most uses of `open type` by definition: qualified access was required before, but `open type` enables unqualified access.

For what it's worth, `using static` on an enum in C# does bring the enum fields into scope both as expressions and patterns.

**Comment by auduchinok** [comment]
> But that is true of most uses of open type by definition: qualified access was required before, but open type enables unqualified access.

IIRC `open type` was only designed to bring static members like methods and properties in scope. It should not affect things like union case or record fields when the type has `RequireQualifiedAccess` attribute.

```fsharp
module Module

[<RequireQualifiedAccess>]
type U =
    | A
    | B

    static member P = 1

open type U

P |> ignore

let a = A
```

<img width="454" alt="Screenshot 2024-09-27 at 17 54 09" src="https://github.com/user-attachments/assets/00aeeb9c-76f2-43af-b4a8-0d7cebefd7b1">

And without the attribute:

<img width="413" alt="Screenshot 2024-09-27 at 17 54 17" src="https://github.com/user-attachments/assets/e2eae693-3782-4f40-95d7-12dd73b4d04d">

**Comment by brianrourkeboll** [comment]
> It should not affect things like union case or record fields when the type has `RequireQualifiedAccess` attribute.

Right. But enums don't have that attribute.

> IIRC `open type` was only designed to bring static members like methods and properties in scope.

It was updated to apply to unions as well (except of course those with `RequireQualifiedAccess`) (https://github.com/fsharp/fslang-design/discussions/352#discussioncomment-231077):

> Resolution: Union cases and record field labels should count as static content when processing `open type` and thus brought into scope..

@abelbraaksma did question the application to enums here: https://github.com/fsharp/fslang-design/discussions/352#discussioncomment-231082. But it doesn't look like anything was recorded in the RFC itself about enums specifically.

**Comment by auduchinok** [comment]
> Right. But enums don't have that attribute.

They do, as per 8.9 in the language spec.

**Comment by brianrourkeboll** [comment]
> They do, as per 8.9 in the language [spec](https://fsharp.org/specs/language-spec/4.1/FSharpSpec-4.1-latest.pdf):

Oh, interesting.

§ [8.9](https://fsharp.org/specs/language-spec/4.1/FSharpSpec-4.1-latest.pdf#page=148)

> Each enum type declaration is implicitly annotated with the `RequiresQualifiedAccess` attribute

The compiler [doesn't actually emit the `RequireQualifiedAccess` attribute for F#-defined enums](https://sharplab.io/#v2:DYLgZgzgPgLgngBwKYAICiKC8BYAUCglKFAQSxQEY9CiUAhcgJiA), though.

Does that really mean "Each enum type declaration [_behaves as though it were_] implicitly annotated with the `RequiresQualifiedAccess` attribute"? And were C#-defined enums meant to be included?

The funny thing is that I would seldom match on F#-defined enums; I would normally just use a union instead. It would normally be a C#-defined enum that I'd want to pattern-match on, anyway, and potentially open using `open type`. So it really depends how C#-defined enums are meant to be treated.

**Comment by vzarytovskii** [comment]
> The compiler doesn't actually emit the RequireQualifiedAccess attribute for F#-defined enums, though.

Yeah, I think the wording is not accurate. I think it's just implicit behaviour for enums rather than implicitly emitting it
